source/darwin/dispatch_queue_event_loop.c

```diff
@@ -0,0 +1,686 @@
+    .free_io_event_resources = s_free_io_event_resources,
+    .is_on_callers_thread = s_is_on_callers_thread,
+};
+
+/**
+ * DISPATCH QUEUE
+ *
+ * Event loop is responsible for processing events and tasks by launching an execution loop on a single thread. Each
+ * iteration of this loop performs three primary jobs:
+ * 1. Process I/O events.
+ * 2. Process cross-thread tasks.
+ * 3. Execute all runnable tasks.
+ *
+ * Apple Dispatch queues are FIFO queues to which the application can submit tasks in the form of block objects, and the
+ * block objects will be executed on a system defined thread pool. Instead of executing the loop on a single thread, we
+ * tried to recurrently run a single iteration of the execution loop as a dispatch queue block object.
+ * aws-c-io library use a sequential dispatch queue to make sure the tasks scheduled on the same dispatch queue are
+ * executed in a strict execution order, though the tasks might be distributed on different threads in the thread pool.
+ *
+ * Data Structures ******
+ * `dispatch_loop_context`: Context for each execution iteration
+ * `scheduled_service_entry`: Each entry maps to each iteration we scheduled on system dispatch queue. As we lost
+ * control of the submitted block on the system dispatch queue, the entry is what we used to track the context and user
+ * data.
+ * `dispatch_loop`: Implementation of the event loop for dispatch queue.
+ *
+ * Functions ************
+ * `s_run_iteration`: The function execute on each single iteration
+ * `begin_iteration`: Decide if we should run the iteration
+ * `end_iteration`: Clean up the related resource and determine if we should schedule next iteration
+ *
+ */
+
+/* The dispatch_scheduling_state holds required information to schedule a "block" on the dispatch_queue. */
+struct dispatch_scheduling_state {
+
+    /**
+     * The lock is used to protect the scheduled_services list cross threads. It should be hold while we add/remove
+     * entries from the scheduled_services list.
+     */
+    struct aws_mutex services_lock;
+    /**
+     * priority queue of <scheduled_service_entry> in sorted order by timestamp. Each scheduled_service_entry represents
+     * a block ALREADY SCHEDULED on apple dispatch queue.
+     *
+     * When we go to schedule a new iteration, we check here first to see if our scheduling attempt is redundant.
+     */
+    struct aws_priority_queue scheduled_services;
+};
+
+/* Internal ref-counted dispatch loop context to processing Apple Dispatch Queue Resources */
+struct dispatch_loop_context {
+    /**
+     * The conetxt lock is a read-write lock used to protect dispatch_loop.
+     * The write lock will be acquired when we make changes to dispatch_loop. And the read lock will be acquired
+     * when we need verify if the dispatch_loop is alive. This makes sure that the dispatch_loop will not be destroyed
+     * from other thread while we are using it.
+     */
+    struct aws_rw_lock lock;
```

Comment:
Trivial: The comment above refers to this lock as the "conetxt lock". Spelling on "context lock" and also we should clarify by either calling it "The dispatch_loop_context's lock" or renaming the lock member here to "context_lock"